src/core/service.c

```diff
@@ -3556,36 +3554,11 @@ static int service_dispatch_exec_io(sd_event_source *source, int fd, uint32_t ev
 
                         return 0;
                 }
-                if (n == 1) {
-                        /* Backward compatibility block: a single byte was read, which means somehow an old
-                         * executor before this logic was introduced sent the notification, so there is no
-                         * timestamp (reset it). */
-
-                        s->exec_fd_hot = x;
-                        if (s->state == SERVICE_START)
-                                s->main_exec_status.handover_timestamp = DUAL_TIMESTAMP_NULL;
-
-                        continue;
-                }
 
-                if (x == 0) {
-                        /* If we get x=0 then the execve actually failed, which means the exec fd logic is
-                         * now off. Also reset the exec timestamp, given it has no meaning anymore. */
+                /* A byte was read → this turns on/off the exec fd logic */
+                assert(n == sizeof(x));
 
-                        s->exec_fd_hot = false;
-                        if (s->state == SERVICE_START)
-                                s->main_exec_status.handover_timestamp = DUAL_TIMESTAMP_NULL;
```

Comment:
Thanks for adding the pre-exec failure case - the post-exec failure case that was covered here is still missing though. IE, if sighup comes and the logic was disabled, the metric needs to be reset too

Reply:
but why? i think it's really interesting to know when the handoff was attempted, even if it failed later. why hide the timestamp? 

that the handoff failed is recorded, hence if clients want to ignore the timestamp in that case they can suppress it. But i think it's valuable to collect the handoff attempt time either way.

I mean, this is generally what logging does: it collects not just info about things that failed, but also *when* they failed. Hence, here, for tracking activation results I also think it's highly interesting to know *when* the handoff failed if it failed.

Reply:
If this was a log then of course you'd be right - but it isn't, it's a metric, which is subtly different in that it has a semantic meaning. And the meaning is: we gave this to the kernel at this time, not our job anymore. But in this failure case that _doesn't happen_, so it's just confusing to have the metric be available for an event that did not actually happen, and it remains our job to handle it - with retries or with failure reporting depending on configuration.

Having a log message in addition, that is always reported regardless, would be of course good for me to have, on top of the separate metric. Then, you always get the reporting of the attempt with the messaged timestamp in the logs, but the metric's meaning remains the same.

Reply:
Sorry, but i just disagree. The data is relevant either way. There's really no need to suppress it.

Reply:
It's not relevant for the actual use case, as it changes the meaning to something different, as explained. How about adding a log message, so that you get the data, and resetting the metric, so that it keeps the meaning it was added for? Best of both worlds and we can move on.

Reply:
I think it's better to take the timestamp in all cases. The timestamp is a piece of information that is independent of whether the exec succeeded.

Reply:
Taken and sent back and logged sure, but I'd like it to be reset in the public API, as in, the new dbus property, for the reasons already explained

Reply:
@poettering this is still outstanding, which I gather means you have no intention of fixing as asked. That's fine, but could you please at least document the behaviour in `man/org.freedesktop.systemd1.xml` so that there are no surprises?

Then I'll approve. Thanks.